### PR TITLE
fix(one-time backup): fix bug when backup request is timeout

### PR DESCRIPTION
**Pull request by Smityz**
https://github.com/apache/incubator-pegasus/issues/893

ref https://github.com/XiaoMi/rdsn/blob/master/src/meta/backup_engine.cpp#L202
If `err`  == TIME_OUT, `response` will be empty, then coredump.

Tests 
- fix the origin unit test
- manual test in onebox

### CODE REVIEW

File: src/meta/backup_engine.cpp [hunk 1]
```diff
@@ -215,37 +212,53 @@ void backup_engine::on_backup_reply(error_code err,
...
+                     err.to_string(),
+                     response.err.to_string());
+            zauto_lock l(_lock);
+            // if one partition fail, the whole backup plan fail.
+            _is_backup_failed = true;
+            _backup_status[pid.get_partition_index()] = backup_status::FAILED;
+            return;
+        } else {
+            // backup not completed, retry sending requests
+            dwarn_f("backup_id({}): send backup request to server {} failed, rpc error: {}, "
```
Comment (levy5307) [inline src/meta/backup_engine.cpp:233]: These code is same with L261~L275, could you refactor it?

Comment (Smityz) [reply src/meta/backup_engine.cpp:233]: It's not the same

Comment (levy5307) [reply src/meta/backup_engine.cpp:233]: I found there are a lot of same code between them

Comment (Smityz) [reply src/meta/backup_engine.cpp:233]: the logic is different

File: src/meta/backup_engine.cpp [hunk 2]
```diff
@@ -215,51 +212,69 @@ void backup_engine::on_backup_reply(error_code err,
     // if backup failed, receive ERR_LOCAL_APP_FAILURE;
     // backup not completed in other cases.
     // see replica::on_cold_backup() for details.
-    int32_t partition = pid.get_partition_index();
-    if (err == dsn::ERR_OK && response.err == dsn::ERR_OK &&
-        response.progress == cold_backup_constant::PROGRESS_FINISHED) {
+    if (err != ERR_OK || response.err != ERR_OK) {
```
Comment (hycdong) [inline src/meta/backup_engine.cpp:215]: I recommend refactor your `on_backup_reply` function like following:
```
if(err != ERR_OK)
{
    dwarn_f("backup_id({}): send backup request to server {} failed, rpc error: {}, retry to "
                "send backup request.",
                _cur_backup.backup_id,
                primary.to_string(),
                err.to_string());
    resend_backup_app_partition_after_1_seconds;
    return;
}

// The following code handle err == ERR_OK (rpc succeed)
dcheck_eq(response.pid, pid);
dcheck_eq(response.backup_id, _cur_backup.backup_id);
{
      zauto_lock l(_lock);
      // if backup of some partition failed, we would not handle response from other partitions.
      if (_is_backup_failed) {
          return;
      }
  }

// backup failed case
if (response.err == ERR_LOCAL_APP_FAILURE) {
        derror_f("backup_id({}): backup for partition {} failed.",
                 _cur_backup.backup_id,
                 pid.to_string());
        zauto_lock l(_lock);
        _is_backup_failed = true;
        _backup_status[partition] = backup_status::FAILED;
        return;
    }

// backup finish case
if (response.err == dsn::ERR_OK && response.progress == cold_backup_constant::PROGRESS_FINISHED) {
        ddebug_f("backup_id({}): backup for partition {} completed.",
                 _cur_backup.backup_id,
                 pid.to_string());
        {
            zauto_lock l(_lock);
            _backup_status[partition] = backup_status::COMPLETED;
        }
        complete_current_backup();
        return;
    }

// other retry cases
// backup is not finished, meta polling to send request
        ddebug_f(
            "backup_id({}): receive backup response for partition {} from server {}, rpc error "
            "{}, response error {}, now progress {}, retry to send backup request.",
            _cur_backup.backup_id,
            pid.to_string(),
            primary.to_string(),
            err.to_string(),
            response.err.to_string(),
            response.progress);

        tasking::enqueue(LPC_DEFAULT_CALLBACK,
                         &_tracker,
                         [this, pid]() { backup_app_partition(pid); },
                         0,
                         std::chrono::seconds(1));
```

File: src/meta/backup_engine.cpp [hunk 3]
```diff
@@ -215,51 +241,63 @@ void backup_engine::on_backup_reply(error_code err,
...
-    if (err == dsn::ERR_OK && response.err == dsn::ERR_OK &&
-        response.progress == cold_backup_constant::PROGRESS_FINISHED) {
+    if (err != ERR_OK) {
+        dwarn_f("backup_id({}): send backup request to server {} failed, rpc error: {}, "
+                "response.err: {} , retry to "
+                "send backup request.",
+                _cur_backup.backup_id,
+                primary.to_string(),
+                err.to_string(),
+                response.err.to_string());
```
Comment (hycdong) [inline src/meta/backup_engine.cpp:251]: I recommend using `derror_f`, and removing response.err in this log.

File: src/meta/backup_engine.cpp [hunk 4]
```diff
@@ -194,14 +194,40 @@ void backup_engine::backup_app_partition(const gpid &pid)
     _backup_status[pid.get_partition_index()] = backup_status::ALIVE;
 }
 
-void backup_engine::on_backup_reply(error_code err,
-                                    const backup_response &response,
-                                    gpid pid,
-                                    const rpc_address &primary)
+inline void backup_engine::handle_replica_backup_failed(const error_code err,
+                                                        const backup_response &response,
+                                                        const gpid pid)
```
Comment (hycdong) [inline src/meta/backup_engine.cpp:199]: It seems that `err` is useless in this function.

File: src/meta/backup_engine.cpp [hunk 5]
```diff
@@ -215,51 +241,63 @@ void backup_engine::on_backup_reply(error_code err,
...
+    }
+
+    if (response.err != ERR_OK) {
+        dwarn_f("backup_id({}): replica {} backup failed, rpc error: {}, "
+                "response.err: {} , retry to "
+                "send backup request.",
+                _cur_backup.backup_id,
+                primary.to_string(),
+                err.to_string(),
+                response.err.to_string());
```
Comment (hycdong) [inline src/meta/backup_engine.cpp:269]: Same, recommend using derror_f, and removing err in this log.

File: src/meta/backup_engine.cpp [hunk 6]
```diff
@@ -215,51 +241,63 @@ void backup_engine::on_backup_reply(error_code err,
...
-    }
+    // backup is not finished, meta polling to send request
+    ddebug_f("backup_id({}): receive backup response for partition {} from server {}, rpc error "
+             "{}, response error {}, now progress {}, retry to send backup request.",
+             _cur_backup.backup_id,
+             pid.to_string(),
+             primary.to_string(),
+             err.to_string(),
+             response.err.to_string(),
+             response.progress);
```
Comment (hycdong) [inline src/meta/backup_engine.cpp:297]: removing err in this log.

File: src/meta/backup_engine.cpp [hunk 7]
```diff
@@ -194,14 +194,37 @@ void backup_engine::backup_app_partition(const gpid &pid)
...
 
+    derror_f("backup_id({}): backup for partition {} failed, response.err: {}",
+             _cur_backup.backup_id,
+             pid.to_string(),
+             response.err.to_string());
+    zauto_lock l(_lock);
+    // if one partition fail, the whole backup plan fail.
+    _is_backup_failed = true;
+    _backup_status[pid.get_partition_index()] = backup_status::FAILED;
+    return;
```
Comment (hycdong) [inline src/meta/backup_engine.cpp:211]: This `return` can be removed.

File: src/meta/backup_engine.cpp [hunk 8]
```diff
@@ -215,51 +238,57 @@ void backup_engine::on_backup_reply(error_code err,
...
-            err.to_string(),
-            response.err.to_string());
-    }
-    tasking::enqueue(LPC_DEFAULT_CALLBACK,
-                     &_tracker,
-                     [this, pid]() { backup_app_partition(pid); },
-                     0,
-                     std::chrono::seconds(1));
+    retry_backup(pid);
+    return;
```
Comment (hycdong) [inline src/meta/backup_engine.cpp:291]: Same, this `return` can be removed.

**Review by hycdong — APPROVED** [review]

File: src/meta/backup_engine.cpp [hunk 9]
```diff
@@ -215,51 +237,56 @@ void backup_engine::on_backup_reply(error_code err,
     // if backup failed, receive ERR_LOCAL_APP_FAILURE;
     // backup not completed in other cases.
     // see replica::on_cold_backup() for details.
-    int32_t partition = pid.get_partition_index();
-    if (err == dsn::ERR_OK && response.err == dsn::ERR_OK &&
-        response.progress == cold_backup_constant::PROGRESS_FINISHED) {
+    if (err != ERR_OK) {
```
Comment (Shuo-Jia) [inline src/meta/backup_engine.cpp:240]:
```c++
err = err == ERR_OK?  resp.err : err;

if err == ERR_LOCAL_APP_FAILURE {
   handle_replica_backup_failed
   return
}

if (err != ERR_OK) {
  retry.... 
}
```


The you just check `err`,  and don't need repeat L256

File: src/meta/backup_engine.cpp [hunk 10]
```diff
@@ -194,14 +194,36 @@ void backup_engine::backup_app_partition(const gpid &pid)
...
+    _backup_status[pid.get_partition_index()] = backup_status::FAILED;
+}
+
+inline void backup_engine::retry_backup(const dsn::gpid pid)
+{
+    tasking::enqueue(LPC_DEFAULT_CALLBACK,
+                     &_tracker,
+                     [this, pid]() { backup_app_partition(pid); },
+                     0,
+                     std::chrono::seconds(1));
```
Comment (Shuo-Jia) [inline src/meta/backup_engine.cpp:219]: consider assign 10s

File: src/meta/backup_engine.cpp (hunk 9 above)
Comment (Smityz) [reply src/meta/backup_engine.cpp:240]: It's better to write easy reading codes than simple codes

File: src/meta/backup_engine.cpp (hunk 10 above)
Comment (Smityz) [reply src/meta/backup_engine.cpp:219]: Why

File: src/meta/backup_engine.cpp (hunk 9 above)
Comment (Shuo-Jia) [reply src/meta/backup_engine.cpp:240]: Your code has been duplicated,  why the new code is not-reading?

File: src/meta/backup_engine.cpp (hunk 10 above)
Comment (Shuo-Jia) [reply src/meta/backup_engine.cpp:219]: it's not necessary to using too frequently value

File: src/meta/backup_engine.cpp (hunk 9 above)
Comment (Smityz) [reply src/meta/backup_engine.cpp:240]: response is a const variable

**Review by Shuo-Jia — DISMISSED** [review]

File: src/meta/backup_engine.cpp [hunk 11]
```diff
@@ -215,51 +237,47 @@ void backup_engine::on_backup_reply(error_code err,
...
+
+    auto rep_error = err == ERR_OK ? response.err : err;
+
+    if (rep_error == ERR_LOCAL_APP_FAILURE) {
+        handle_replica_backup_failed(response, pid);
+        return;
+    }
+
+    if (rep_error != ERR_OK) {
+        derror_f("backup_id({}): backup request to server {} failed, rpc error: {}, retry to "
```
Comment (hycdong) [inline src/meta/backup_engine.cpp:249]: "rpc error" -> "error"

Comment (Smityz) [inline src/meta/backup_engine.cpp:249]:
```suggestion
        derror_f("backup_id({}): backup request to server {} failed, error: {}, retry to "
```

**Review by hycdong — APPROVED** [review]

**Review by Shuo-Jia — APPROVED** [review]

**Pull request merged**